scripts/1.TCGA-MLexample.py

```diff
@@ -72,84 +75,96 @@
 
 
-# In[8]:
+# In[9]:
 
 get_ipython().run_cell_magic('time', '', "path = os.path.join('data', 'mutation-matrix.tsv.bz2')\nY = pd.read_table(path, index_col=0)")
 
 
-# ## Process Mutation Matrix
-
-# In[9]:
+# In[10]:
 
 y = Y[GENE]
 
 
-# In[10]:
+# In[11]:
 
 # The Series now holds TP53 Mutation Status for each Sample
 y.head(6)
 
 
-# In[11]:
+# In[12]:
 
 # Here are the percentage of tumors with NF1
 y.value_counts(True)
 
 
 # ## Set aside 10% of the data for testing
 
-# In[12]:
+# In[13]:
 
 # Typically, this can only be done where the number of mutations is large enough
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
 'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(len(X.index), len(X_train), len(X_test))
 
 
-# ## Feature transformation: Standardization
+# ## Median absolute deviation feature selection
 
-# In[13]:
+# In[14]:
 
-def scale_df(df):
-    """Standardize colummns by scaling to z-scores"""
-    scaled_matrix = preprocessing.scale(df)
-    return pd.DataFrame(scaled_matrix, index=df.index, columns=df.columns)
+def fs_mad(x, y):
+    mad = lambda a: np.median(np.absolute(a - np.median(a)))
+    score = np.apply_along_axis(mad, 0, x)
+    p = np.array([0.5]*len(score))
+    return score, p
+feature_select = SelectKBest(fs_mad, k=500)  # select the top 500 features with the highest MAD
 
-# Standardize X separately for the holdout dataset
-X_train, X_test = scale_df(X_train), scale_df(X_test)
 
+# ## Define pipeline and Cross validation model fitting
 
-# ## Cross validation model fitting
+# In[15]:
 
-# In[14]:
+# Include loss='log' in param_grid doesn't work with pipeline somehow
+clf = SGDClassifier(random_state=0, class_weight='balanced', loss='log')
 
 # joblib is used to cross-validate in parallel by setting `n_jobs=-1` in GridSearchCV
 # Supress joblib warning. See https://github.com/scikit-learn/scikit-learn/issues/6370
 warnings.filterwarnings('ignore', message='Changing the shape of non-C contiguous array')
+clf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')
+pipeline = make_pipeline(
+              feature_select,  # Feature selection
+              StandardScaler(),  # Feature scaling
+              clf_grid)
 
 
-# In[15]:
+# In[16]:
 
-get_ipython().run_cell_magic('time', '', "clf = SGDClassifier(random_state=0, class_weight='balanced')\nclf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')\n# Fit the model (the computationally intensive part)\nclf_grid.fit(X=X_train, y=y_train)\nbest_clf = clf_grid.best_estimator_")
+get_ipython().run_cell_magic('time', '', '# Fit the model (the computationally intensive part)\npipeline.fit(X=X_train, y=y_train)\nbest_clf = clf_grid.best_estimator_')
 
 
-# In[16]:
+# In[17]:
+
+# Check the list of selected features
```

Comment:
I don't find this cell helpful for the final notebook. i.e. great for development but remove once you see it worked as intended? In other words, I don't think the notebook viewer is gaining much from this cell and the output is visually distracting.